src/hamiltonian.jl

```diff
@@ -1,20 +1,23 @@
 # TODO: add a type for kinetic energy
 # TODO: add cache for gradients by letting ∂logπ∂θ return both log density and gradient
 
-struct Hamiltonian{M<:AbstractMetric, Tlogπ, T∂logπ∂θ}
+struct Hamiltonian{M<:AbstractMetric, Tlogπ, TF∂logπ∂θ, T∂H∂θ, T∂H∂r}
     metric::M
     ℓπ::Tlogπ
-    ∂ℓπ∂θ::T∂logπ∂θ
+    ∂ℓπ∂θ!::TF∂logπ∂θ
+    ∂H∂θ::T∂H∂θ
```

Comment:
I would suggest that we remove these caching in `Hamoltonian`. `∂H∂θ` and `∂H∂r` are already cached by `PhasePoint`.  These caching are better done in `PhasePoint` instead of `Hamiltonian` since they are phasepoint-specific.